Feat: 스터디룸에 채팅 기능 추가

OHJINSEONG opened the pull request:
- 채팅 모델 구현
- 스터디룸 입장시 채팅 데이터 조회
- 채팅 입력시 실시간 업데이트

OHJINSEONG committed:
Feat: 스터디룸에 채팅 기능 추가

- 채팅 모델 구현
- 스터디룸 입장시 채팅 데이터 조회
- 채팅 입력시 실시간 업데이트

junhyeong00 commented on src/main/java/com/pomodoro/pomodoromate/chat/controllers/MessageController.java:
```diff
@@ -0,0 +1,44 @@
+
+        ChatSummariesDto chatSummariesDto = chatMessageService.findChats(studyRoomId);
+
+        template.convertAndSend("/sub/user/chat", chatSummariesDto);
```
메시지 전달을 컨트롤러단에서 하는 이유가 있나요?

OHJINSEONG replied:
서비스단에 메세지 전달을 하면 서비스의 메서드 유지보수성이나 확장성이 오히려 떨어진다고 생각해서 컨트롤 단으로 빼버렸습니다. 그런데 생각해보니 서비스 메서드를 save() -> chat() 이런 식으로 바꾸고 메세지 전달까지 위임하면 더 좋겠네요!

junhyeong00 commented on src/main/java/com/pomodoro/pomodoromate/chat/controllers/MessageController.java:
```diff
@@ -0,0 +1,44 @@
+public class MessageController {
+    private final SimpMessagingTemplate template;
+    private final ChatMessageService chatMessageService;
+    private final JwtUtil jwtUtil;
```
jwtUtil은 어디서 사용되고 있는건가요?

OHJINSEONG replied:
토큰받은거 인가 목적으로 사용했다가 추후에 추가하려고 지웠는데 jwtUtil은 안지웠내요 ㅠㅠ

junhyeong00 commented on src/main/java/com/pomodoro/pomodoromate/chat/controllers/MessageController.java:
```diff
@@ -0,0 +1,44 @@
+    private final ChatMessageService chatMessageService;
+    private final JwtUtil jwtUtil;
+
+    public MessageController(SimpMessagingTemplate template, ChatMessageService chatMessageService, JwtUtil jwtUtil) {
```
코드가 길어서 줄바꿈 해주는 게 좋을 것 같아요~

OHJINSEONG replied:
넵

junhyeong00 commented on src/main/java/com/pomodoro/pomodoromate/chat/applications/ChatMessageService.java:
```diff
@@ -0,0 +1,49 @@
+import java.util.List;
+
+@Service
+@Transactional
```
Transactional을 메서드가 아니라 클래스에 붙인 이유가 있나요?

OHJINSEONG replied:
모든 메서드 전부 사용해서 일단 클래스에 붙였습니다

junhyeong00 replied:
find 메서드는 읽기만 해서 클래스가 아닌 메서드에서 readonly 옵션 붙여주는 게 더 좋을 것 같아요!

junhyeong00 commented on src/main/java/com/pomodoro/pomodoromate/chat/models/Chat.java:
```diff
@@ -0,0 +1,53 @@
+@Entity
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+public class Chat extends BaseEntity {
+    @Id
+    @GeneratedValue
+    private Long id;
+
+    @Embedded
+    @AttributeOverride(name = "value", column = @Column(name = "studyRoomId"))
+    private StudyRoomId studyRoomId;
+
+    @Embedded
+    @AttributeOverride(name = "value", column = @Column(name = "participantId"))
+    private ParticipantId participantId;
+
+    @Embedded
+    private Message message;
+
+    @Embedded
+    private Writer writer;
+
+    @Builder
+    public Chat(Long id, StudyRoomId studyRoomId, ParticipantId participantId, Message message, Writer writer) {
+        this.id = id;
+        this.studyRoomId = studyRoomId;
+        this.participantId = participantId;
+        this.message = message;
+        this.writer = writer;
+    }
```
이미지 채팅과 일반 채팅은 어떻게 구별하실 생각이신가요?

OHJINSEONG replied:
이미지 채팅이라 함은 어떤것 말씀하시는거죠? 이미지를 채팅에 올리는 것 말씀하시는 건가요?

junhyeong00 replied:
네 맞습니다!

OHJINSEONG committed:
Chore: 웹소캣 채팅 전달 로직을 서비스 레이어로 리펙토링

- 컨트롤러 레이어에 있던 웹소캣 메세지 전달 로직을 서비스 레이어로 이동
- GetChatMessageService를 만들어 책임 분리
- 채팅 서비스 레이어 메서드 이름 변경

junhyeong00 commented on src/main/java/com/pomodoro/pomodoromate/chat/controllers/MessageController.java:
```diff
@@ -0,0 +1,38 @@
+    @MessageMapping("/user/chat/enter")
+    public void userEnter(
+    ) {
+        StudyRoomId studyRoomId = new StudyRoomId(1L);
+
```
스터디룸 id가 1고정인가요..?

OHJINSEONG replied:
임시로 만들려다 보니 실수했네요!

junhyeong00 commented on src/main/java/com/pomodoro/pomodoromate/chat/applications/ChatMessageService.java:
```diff
@@ -0,0 +1,45 @@
+
+        ChatSummaryDto chatSummaryDto = saved.toSummaryDto();
+
+        template.convertAndSend("/sub/chat/room/" + chatSummaryDto.roomId(), chatSummaryDto);
+    }
```
/sub/studyrooms/" + studyRoomId + "/chat" 같은 느낌은 어떤가요?

OHJINSEONG replied:
좋습니다!

junhyeong00 commented on src/main/java/com/pomodoro/pomodoromate/chat/applications/GetChatMessageService.java:
```diff
@@ -0,0 +1,39 @@
+        List<ChatSummaryDto> chatSummaryDtos = chats.stream()
+                .map(Chat::toSummaryDto).toList();
+
+        template.convertAndSend("/sub/user/chat", new ChatSummariesDto(chatSummaryDtos));
```
/sub/user/chat 이거는 무슨의미인가요?

OHJINSEONG replied:
다시 읽어보니 이상하네요 수정하겠습니다!

junhyeong00 commented on src/main/java/com/pomodoro/pomodoromate/chat/controllers/MessageController.java:
```diff
@@ -0,0 +1,38 @@
+        this.chatMessageService = chatMessageService;
+    }
+
+    @MessageMapping("/user/chat/enter")
+    public void userEnter(
+    ) {
+        StudyRoomId studyRoomId = new StudyRoomId(1L);
+
+        getChatMessageService.getChatsForEntry(studyRoomId);
+    }
```
입장했을 때, 채팅을 불러오는 것보다 스터디룸 조회를 했을 때, 가져오는 게 좋지 않을까요?

OHJINSEONG replied:
그렇게 수정하겠습니다!

junhyeong00 commented on src/main/java/com/pomodoro/pomodoromate/chat/controllers/MessageController.java:
```diff
@@ -0,0 +1,38 @@
+    ) {
+        StudyRoomId studyRoomId = new StudyRoomId(1L);
+
+        getChatMessageService.getChatsForEntry(studyRoomId);
+    }
```
getChatsForEntry는 어떤 의미인가요? 입장하기 위한 채팅을 가져온다는 의미인가요?

OHJINSEONG replied:
넵!!

junhyeong00 commented on src/main/java/com/pomodoro/pomodoromate/chat/controllers/MessageController.java:
```diff
@@ -0,0 +1,38 @@
+
+
+    @MessageMapping("/user/chat/message")
+    public void message(
+            ChatRequestDto chatRequestDto
```
메서드명은 동사형태가 좋을 것 같아요~ 예를들면 sendMessage처럼요

OHJINSEONG replied:
좋습니다!

junhyeong00 commented on src/main/java/com/pomodoro/pomodoromate/chat/applications/GetChatMessageService.java:
```diff
@@ -0,0 +1,39 @@
+        List<ChatSummaryDto> chatSummaryDtos = chats.stream()
+                .map(Chat::toSummaryDto).toList();
+
+        template.convertAndSend("/sub/user/chat", new ChatSummariesDto(chatSummaryDtos));
```
채팅 내용을 웹소켓으로 전달할 필요가 있나요?

OHJINSEONG replied:
채팅 내역을 가져오고 웹소캣 연결시 누락되는 문제가 발생할 수 있다고 들어서 웹소캣으로 전달 했는데 웹소캣 연결후 데이터 가져오고 초기 데이터와 웹소켓 메시지 동기화 작업도 추가해보겠습니다

junhyeong00 commented on src/main/java/com/pomodoro/pomodoromate/chat/applications/ChatMessageService.java:
```diff
@@ -0,0 +1,45 @@
+                .message(Message.of(chatRequestDto.message()))
+                .build();
+
+        Chat saved = chatMessageRepository.save(chat);
```
클래스명을 chat으로 할거면 레포지토리도 chatRepository로 하는 게 어떨까요?

OHJINSEONG replied:
좋습니다!

OHJINSEONG committed:
Chore: 채팅 목록 데이터 수신 웹소캣에서 HTTP로 수정

- 채팅 목록을 기존의 WebSocket을 통해 실시간으로 받아오던 방식에서 HTTP 요청을 통해 DTO로 받아오는 방식으로 변경.
- 스터디룸 정보 불러올시 참가자 DTO 변환 로직 서비스 레이어에서 도메인 레이어로 교체

junhyeong00 commented on src/main/java/com/pomodoro/pomodoromate/chat/applications/ChatMessageService.java:
```diff
@@ -0,0 +1,42 @@
+        this.template = template;
+    }
+
+    @Transactional
+    public void sendMessage(ChatRequestDto chatRequestDto) {
+        Chat chat = Chat.builder()
+                .studyRoomId(StudyRoomId.of(chatRequestDto.studyRoomId()))
+                .participantId(ParticipantId.of(chatRequestDto.participantId()))
+                .writer(Writer.of(chatRequestDto.writer()))
+                .message(Message.of(chatRequestDto.message()))
+                .build();
+
+        Chat saved = chatRepository.save(chat);
+
+        ChatSummaryDto chatSummaryDto = saved.toSummaryDto();
+
+        template.convertAndSend("/sub/studyRooms/" + chatSummaryDto.roomId() + "/chat", chatSummaryDto);
+    }
```
참가자(현재 해당 스터디룸에서 참가 상태인지)와 사용자(인증된 사용자가 맞는지) 검증 로직이 필요할 것 같아요!

OHJINSEONG replied:
넵 알겠습니다!

junhyeong00 commented on src/main/java/com/pomodoro/pomodoromate/studyRoom/applications/GetStudyRoomService.java:
```diff
@@ -39,12 +43,8 @@ public StudyRoomDetailDto studyRoom(Long studyRoomId, UserId userId) {
 
-        List<ParticipantSummaryDto> participantSummaryDtos = participants.stream()
-                .map(participant -> participant.isHost(studyRoom.hostId().value())
-                        ? participant.toSummaryDto(true)
-                        : participant.toSummaryDto(false))
-        .toList();
+        List<Chat> chats = chatRepository.findAllBy(StudyRoomId.of(studyRoomId));
 
-        return studyRoom.toDetailDto(participantSummaryDtos);
+        return studyRoom.toDetailDto(participants, chats);
```
좋은 것 같습니다! 👍

junhyeong00 commented on src/main/java/com/pomodoro/pomodoromate/chat/dtos/ChatRequestDto.java:
```diff
@@ -0,0 +1,11 @@
+
+public record ChatRequestDto (
+        @NotBlank String message,
+        @NotBlank String writer,
```
굳이 작성자 이름을 받아올 필요없이 참가자 검증하면서 저장된 참가자 이름을 가져오면 될 것 같아요!

OHJINSEONG replied:
넵!
